### PR TITLE
Add 'buf format'

### PR DESCRIPTION
**Pull request by amckinney**
This doesn't work, but putting this up as-is since I don't have time to finish right now. This adds a [`DocumentFormattingEditProvider`](https://code.visualstudio.com/api/references/vscode-api#DocumentFormattingEditProvider) that adds `buf format` to the extension. 

A lot of the code here is referenced from Go's old [`goFormat.ts`](https://github.com/golang/vscode-go/blob/0fa41b1c6e4217cc68a61bc97a7cabd34655e3cb/src/language/legacy/goFormat.ts).

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -124,6 +125,7 @@ export function activate(context: vscode.ExtensionContext) {
     diagnosticCollection.set(document.uri, diagnostics);
   };
 
+  context.subscriptions.push(vscode.languages.registerDocumentFormattingEditProvider('proto', new Formatter(binaryPath)));
```
Comment (amckinney) [inline src/extension.ts:128]: I'm not sure that this is right, but it's similar to what we see for Go [here](https://github.com/golang/vscode-go/blob/0fa41b1c6e4217cc68a61bc97a7cabd34655e3cb/src/language/registerDefaultProviders.ts#L45).

File: src/formatter.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        token: vscode.CancellationToken
+    ): Thenable<vscode.TextEdit[]> {
+        return new Promise<vscode.TextEdit[]>((resolve, reject) => {
+            const cwd = path.dirname(document.fileName);
+            let stdout = '';
+            let stderr = '';
+
+            // Use spawn instead of exec to avoid maxBufferExceeded error
+            const p = cp.spawn(this.binaryPath, [document.fileName], { cwd });
+            // TODO: We need to use the CancellationToken.
```
Comment (amckinney) [inline src/formatter.ts:45]: See [this](https://github.com/golang/vscode-go/blob/0fa41b1c6e4217cc68a61bc97a7cabd34655e3cb/src/language/legacy/goFormat.ts#L87).

File: src/formatter.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        return (Math.floor(Math.random() * 100000)).toString()
+    }
+
+    private runFormatter(
+        document: vscode.TextDocument,
+        token: vscode.CancellationToken
+    ): Thenable<vscode.TextEdit[]> {
+        return new Promise<vscode.TextEdit[]>((resolve, reject) => {
+            const cwd = path.join(os.tmpdir(), this.randomId());
+            vscode.workspace.fs.createDirectory(vscode.Uri.file(cwd)).then(() => {
```
Comment (amckinney) [inline src/formatter.ts:45]: I know it's extremely unlikely (and probably overkill), but should we check if the directory exists before we proceed with the random ID we selected? We could continue to generate a new ID until we find one that isn't already occupied (stopping after a certain number of attempts so we don't try forever). This is similar to what [os.CreateTemp](https://cs.opensource.google/go/go/+/refs/tags/go1.18:src/os/tempfile.go;drc=refs%2Ftags%2Fgo1.18;l=30) does.

It seems like a simple enough function to build out, so it might be worth it.

File: src/formatter.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            (edits) => edits,
+            (err) => {
+                console.log(err);
+                return Promise.reject('Check the console in dev tools to find errors when formatting.');
+            }
+        );
+    }
+
+    private randomId(): string {
+        return (Math.floor(Math.random() * 100000)).toString()
```
[the remark this answers is not in the document]
Comment (amckinney) [reply src/formatter.ts:36]: Don't forget to fix lint here (the missing `;`).

File: package.json [hunk 5]
```diff
@@ -70,6 +77,16 @@
...
 				]
+			},
+			{
+				"id": "proto",
+				"extensions": [
+					".proto"
+				],
+				"aliases": [
+					"Protocol Buffers",
+					"Protobufs"
```
Comment (amckinney) [inline package.json:88]: If we think that the `aliases` are necessary, I'd probably shorten this to "Protobuf" - I always thought that "Protobufs" is a bit of a misnomer.

File: package.json [hunk 6]
```diff
@@ -31,6 +32,7 @@
 		"protobuf",
 		"protocol buffers",
 		"buf",
+		"bufbuild",
 		"lint"
```
Comment (amckinney) [inline package.json:36]: If `lint` is here, let's add `format`, too.

**Review by amckinney — COMMENTED** [review]
Looks great - just a few comments, but otherwise looks nearly ready to go.

**Comment by amckinney** [comment]
If we add support for `stdin` (described [here](https://github.com/bufbuild/buf/issues/1035)), a lot of the comments about generating a random ID for temporary directories is a lesser concern - all that will be removed as soon as we have support for stdin.

File: package.json (hunk 6 above)
Comment (rubensf) [reply package.json]: done

File: package.json (hunk 5 above)
Comment (rubensf) [reply package.json]: I don't think they're necessary - IIUC this just improves discoverability (and isn't strictly related to this PR). Changed it to `Protobuf`, but lmk if you'd rather me remove it.

File: src/formatter.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (rubensf) [reply src/formatter.ts]: Yup, the multiplication is just so that we get a whole instead of a float (that and the `Math.floor`). the `floor` guarantees this will be a whole number :)

I at first tried getting some sort of uuid, and looked at some diff libraries but they all had diff issues
* [unique-file-name](https://www.npmjs.com/package/unique-file-name) and [unique-filename](https://www.npmjs.com/package/unique-filename) don't have ts support
* [uniquefilename](https://www.npmjs.com/package/uniquefilename) only does sequential numbers
* [nanoid](https://www.npmjs.com/package/nanoid) just doesn't work? I get some errors when loading the extension "failed to load nanoid" despite it being stalled both globally and locally :(

I figured `100...0` was enough zeroes, so it wasn't worth it spending anymore time learning js. But I added a couple more zeros just in case.

File: src/formatter.ts (hunk 3 above)
Comment (rubensf) [reply src/formatter.ts]: Fair point - I'd _try_ that, but my js-foo is way too subpar to do it cleanly. Given that the chances of collision are tiny, it's likely not worth it (particularly if reading from stdin is coming anytime soon)

Comment (amckinney) [reply src/formatter.ts]: Agreed, I'm fine with leaving this as-is.

File: package.json (hunk 5 above)
Comment (amckinney) [reply package.json]: Sweet, leaving them in seems fine - better discoverability is a good thing.

**Comment by amckinney** [comment]
Looks great - as long as you've verified this all works locally (i.e. formatting on save).

I can't stamp because I'm the original author, but feel free to approve on your end (as a formality on my behalf) and we can merge from there. After this lands, we can create an issue here that says to adapt the implementation to using stdin with that feature exists.

**Review by rubensf — APPROVED** [review]
self-approving, w00t!
